sources/packages/backend/apps/workers/src/services/disbursement-schedule/disbursement-schedule-service.ts

```diff
@@ -1,28 +1,86 @@
+  DisbursementOverawardOriginType,
+  Student,
+  Application,
+  configureIdleTransactionSessionTimeout,
 } from "@sims/sims-db";
-import { Disbursement } from "./disbursement-schedule.models";
+import {
+  DisbursementSaveModel,
+  DisbursementScheduleValue,
+} from "./disbursement-schedule.models";
 import { CustomNamedError } from "@sims/utilities";
 import {
   ASSESSMENT_INVALID_OPERATION_IN_THE_CURRENT_STATE,
   ASSESSMENT_NOT_FOUND,
   DISBURSEMENT_SCHEDULES_ALREADY_CREATED,
 } from "../../constants";
+import { QueryDeepPartialEntity } from "typeorm/query-builder/QueryPartialEntity";
+import { DisbursementOverawardService } from "..";
+import { SystemUsersService } from "@sims/services/system-users";
+
+// Timeout to handle the worst-case scenario where the commit/rollback
+// was not executed due to a possible catastrophic failure.
+const TRANSACTION_IDLE_TIMEOUT_SECONDS = 600;
+
+const LOAN_TYPES = [
+  DisbursementValueType.CanadaLoan,
+  DisbursementValueType.BCLoan,
+];
+
+const GRANTS_TYPES = [
+  DisbursementValueType.CanadaGrant,
+  DisbursementValueType.BCGrant,
+];
 
 /**
  * Service layer for Student Application disbursement schedules.
+ * Assumptions and concepts:
+ * - sims.disbursement_overawards can also be referred as overaward balance table or overaward history table.
+ * - One assessment/reassessment will have one or two dates to be disbursed, a.k.a. disbursement schedule.
+ * - Awards are the loans/grants associated with one disbursement schedule (e.g. CSLF, CSPT, BCSL).
+ * - The same award can be present in one or two schedules.
+ * - A student debt can be a value already received by the student for the same application or prior application.
+ *     - Values already paid(sent/disbursed): sum of all the awards already disbursed for that application number grouped by its code value (e.g. CSLF, CSPT, BCSL).
+ *     - Overaward balance: value from a prior application that must be deducted from the awards, if possible.
+ *     - During a reassessment, the new award calculated will be deducted from all the values already paid for that particular
+ * application. If any award is still to be received, the process will check for possible overaward balance to be deducted.
+ * - If there is a student debt, the priority is to deduct the value as much as possible from the sooner disbursement schedule.
+ * This means that, in case there are two disbursements for the same award, the process will try to deduct as much as possible
+ * from the first disbursement.
+ * - All values once paid(sent/disbursed) for any application associated with an application number (independent of the status)
+ * must be considered as paid.
+ * - Upon reassessment, awards in pending state will have their possible overaward deduction rolled back (added back to the overaward balance table).
+ * - Upon reassessment, if the previous assessment of the same application caused an overaward, this overaward will be soft deleted from the overaward balance table.
```

Comment:
`previous assessment of the same application caused an overaward,`  or `previous assessment of the same application deducted an overward?

Reply:
The first one is correct, as far as I can see. The previous assessment of the same application can generate an overaward that will be added to the overaward balance and will not be deducted at this time because there is no more money left to take from the student.

Reply:
i got it after our discussion